File: sequential-1770261425547/baseline-workspace/payments-ts/src/providers/stripe.ts
import Stripe from 'stripe';
import { PaymentProvider } from '../provider';
import { PaymentError, ErrorCode } from '../errors';
import {
  Charge,
  ChargeRequest,
  Refund,
  RefundRequest,
  Customer,
  CreateCustomerRequest,
  UpdateCustomerRequest,
  PaymentMethod,
  Subscription,
  CreateSubscriptionRequest,
  UpdateSubscriptionRequest,
  CancelSubscriptionRequest,
  WebhookEvent,
  ChargeStatus,
  RefundStatus,
  SubscriptionStatus,
  WebhookEventType,
} from '../types';

export interface StripeConfig {
  apiKey: string;
  webhookSecret: string;
}

export class StripeProvider implements PaymentProvider {
  readonly name = 'stripe';
  private stripe: Stripe;
  private webhookSecret: string;

  constructor(config: StripeConfig) {
    this.stripe = new Stripe(config.apiKey);
    this.webhookSecret = config.webhookSecret;
  }

  // ============ Charges ============

  async createCharge(request: ChargeRequest): Promise<Charge> {
    try {
      const params: Stripe.ChargeCreateParams = {
        amount: request.amount,
        currency: request.currency,
        description: request.description,
        metadata: request.metadata,
      };

      if (request.customerId) params.customer = request.customerId;

      const options: Stripe.RequestOptions = {};
      if (request.idempotencyKey) options.idempotencyKey = request.idempotencyKey;

      const charge = await this.stripe.charges.create(params, options);
      return this.mapCharge(charge);
    } catch (err) {
      throw this.mapError(err);
    }
  }

  async getCharge(chargeId: string): Promise<Charge> {
    try {
      const charge = await this.stripe.charges.retrieve(chargeId);
      return this.mapCharge(charge);
    } catch (err) {
      throw this.mapError(err);
    }
  }

  // ============ Refunds ============

  async createRefund(request: RefundRequest): Promise<Refund> {
    try {
      const params: Stripe.RefundCreateParams = {
        charge: request.chargeId,
      };

      if (request.amount) params.amount = request.amount;
      if (request.reason) params.reason = request.reason as Stripe.RefundCreateParams.Reason;

      const options: Stripe.RequestOptions = {};
      if (request.idempotencyKey) options.idempotencyKey = request.idempotencyKey;

      const refund = await this.stripe.refunds.create(params, options);
      return this.mapRefund(refund);
    } catch (err) {
      throw this.mapError(err);
    }
  }

  async getRefund(refundId: string): Promise<Refund> {
    try {
      const refund = await this.stripe.refunds.retrieve(refundId);
      return this.mapRefund(refund);
    } catch (err) {
      throw this.mapError(err);
    }
  }

  // ============ Customers ============

  async createCustomer(request: CreateCustomerRequest): Promise<Customer> {
    try {
      const customer = await this.stripe.customers.create({
        email: request.email,
        name: request.name,
        phone: request.phone,
        metadata: request.metadata,
      });
      return this.mapCustomer(customer);
    } catch (err) {
      throw this.mapError(err);
    }
  }

  async getCustomer(customerId: string): Promise<Customer> {
    try {
      const customer = await this.stripe.customers.retrieve(customerId);
      if (customer.deleted) {
        throw new PaymentError('customer_not_found', 'Customer has been deleted', 'stripe');
      }
      return this.mapCustomer(customer as Stripe.Customer);
    } catch (err) {
      throw this.mapError(err);
    }
  }

  async updateCustomer(customerId: string, request: UpdateCustomerRequest): Promise<Customer> {
    try {
      const customer = await this.stripe.customers.update(customerId, {
        email: request.email,
        name: request.name,
        phone: request.phone,
        metadata: request.metadata,
      });
      return this.mapCustomer(customer);
    } catch (err) {
      throw this.mapError(err);
    }
  }

  async deleteCustomer(customerId: string): Promise<void> {
    try {
      await this.stripe.customers.del(customerId);
    } catch (err) {
      throw this.mapError(err);
    }
  }

  // ============ Payment Methods ============

  async attachPaymentMethod(customerId: string, paymentMethodId: string): Promise<void> {
    try {
      await this.stripe.paymentMethods.attach(paymentMethodId, { customer: customerId });
    } catch (err) {
      throw this.mapError(err);
    }
  }

  async detachPaymentMethod(paymentMethodId: string): Promise<void> {
    try {
      await this.stripe.paymentMethods.detach(paymentMethodId);
    } catch (err) {
      throw this.mapError(err);
    }
  }

  async listPaymentMethods(customerId: string): Promise<PaymentMethod[]> {
    try {
      const methods = await this.stripe.paymentMethods.list({
        customer: customerId,
        type: 'card',
      });
      return methods.data.map((pm) => this.mapPaymentMethod(pm));
    } catch (err) {
      throw this.mapError(err);
    }
  }

  // ============ Subscriptions ============

  async createSubscription(request: CreateSubscriptionRequest): Promise<Subscription> {
    try {
      const params: Stripe.SubscriptionCreateParams = {
        customer: request.customerId,
        items: [{ price: request.priceId }],
        metadata: request.metadata,
      };

      if (request.paymentMethodId) {
        params.default_payment_method = request.paymentMethodId;
      }

      if (request.trialDays) {
        params.trial_period_days = request.trialDays;
      }

      const options: Stripe.RequestOptions = {};
      if (request.idempotencyKey) options.idempotencyKey = request.idempotencyKey;

      const subscription = await this.stripe.subscriptions.create(params, options);
      return this.mapSubscription(subscription);
    } catch (err) {
      throw this.mapError(err);
    }
  }

  async getSubscription(subscriptionId: string): Promise<Subscription> {
    try {
      const subscription = await this.stripe.subscriptions.retrieve(subscriptionId);
      return this.mapSubscription(subscription);
    } catch (err) {
      throw this.mapError(err);
    }
  }

  async updateSubscription(
    subscriptionId: string,
    request: UpdateSubscriptionRequest
  ): Promise<Subscription> {
    try {
      const params: Stripe.SubscriptionUpdateParams = {
        metadata: request.metadata,
      };

      if (request.paymentMethodId) {
        params.default_payment_method = request.paymentMethodId;
      }

      if (request.cancelAtPeriodEnd !== undefined) {
        params.cancel_at_period_end = request.cancelAtPeriodEnd;
      }

      // Plan change requires updating items
      if (request.priceId) {
        const current = await this.stripe.subscriptions.retrieve(subscriptionId);
        if (current.items.data.length > 0) {
          params.items = [
            {
              id: current.items.data[0].id,
              price: request.priceId,
            },
          ];
        }
      }

      const subscription = await this.stripe.subscriptions.update(subscriptionId, params);
      return this.mapSubscription(subscription);
    } catch (err) {
      throw this.mapError(err);
    }
  }

  async cancelSubscription(
    subscriptionId: string,
    request: CancelSubscriptionRequest
  ): Promise<Subscription> {
    try {
      let subscription: Stripe.Subscription;

      if (request.cancelAtPeriodEnd) {
        subscription = await this.stripe.subscriptions.update(subscriptionId, {
          cancel_at_period_end: true,
        });
      } else {
        subscription = await this.stripe.subscriptions.cancel(subscriptionId);
      }

      return this.mapSubscription(subscription);
    } catch (err) {
      throw this.mapError(err);
    }
  }

  async listSubscriptions(customerId: string): Promise<Subscription[]> {
    try {
      const subscriptions = await this.stripe.subscriptions.list({ customer: customerId });
      return subscriptions.data.map((sub) => this.mapSubscription(sub));
    } catch (err) {
      throw this.mapError(err);
    }
  }

  // ============ Webhooks ============

  async parseWebhook(payload: Buffer | string, signature: string): Promise<WebhookEvent> {
    try {
      const event = this.stripe.webhooks.constructEvent(
        payload,
        signature,
        this.webhookSecret
      );

      return {
        id: event.id,
        type: this.mapEventType(event.type),
        provider: 'stripe',
        data: event.data.object,
        createdAt: new Date(event.created * 1000),
      };
    } catch {
      throw new PaymentError('invalid_webhook', 'Invalid webhook signature', 'stripe');
    }
  }

  // ============ Mappers ============

  private mapCharge(charge: Stripe.Charge): Charge {
    return {
      id: charge.id,
      providerId: charge.id,
      provider: 'stripe',
      amount: charge.amount,
      amountRefunded: charge.amount_refunded,
      currency: charge.currency,
      status: this.mapChargeStatus(charge.status),
      customerId: typeof charge.customer === 'string' ? charge.customer : charge.customer?.id,
      description: charge.description ?? undefined,
      failureCode: charge.failure_code ?? undefined,
      failureReason: charge.failure_message ?? undefined,
      metadata: charge.metadata ?? {},
      createdAt: new Date(charge.created * 1000),
    };
  }

  private mapChargeStatus(status: string): ChargeStatus {
    switch (status) {
      case 'succeeded': return 'succeeded';
      case 'pending': return 'pending';
      case 'failed': return 'failed';
      default: return status as ChargeStatus;
    }
  }

  private mapRefund(refund: Stripe.Refund): Refund {
    return {
      id: refund.id,
      providerId: refund.id,
      provider: 'stripe',
      chargeId: typeof refund.charge === 'string' ? refund.charge : refund.charge?.id ?? '',
      amount: refund.amount,
      currency: refund.currency,
      status: this.mapRefundStatus(refund.status),
      reason: refund.reason ?? undefined,
      createdAt: new Date(refund.created * 1000),
    };
  }

  private mapRefundStatus(status: string | null): RefundStatus {
    switch (status) {
      case 'succeeded': return 'succeeded';
      case 'pending': return 'pending';
      case 'failed': return 'failed';
      default: return 'pending';
    }
  }

  private mapCustomer(customer: Stripe.Customer): Customer {
    return {
      id: customer.id,
      providerId: customer.id,
      provider: 'stripe',
      email: customer.email ?? '',
      name: customer.name ?? undefined,
      phone: customer.phone ?? undefined,
      metadata: customer.metadata ?? {},
      createdAt: new Date(customer.created * 1000),
    };
  }

  private mapPaymentMethod(pm: Stripe.PaymentMethod): PaymentMethod {
    return {
      id: pm.id,
      providerId: pm.id,
      provider: 'stripe',
      type: 'card',
      card: pm.card
        ? {
            brand: pm.card.brand,
            last4: pm.card.last4,
            expMonth: pm.card.exp_month,
            expYear: pm.card.exp_year,
          }
        : undefined,
      isDefault: false,
      createdAt: new Date(pm.created * 1000),
    };
  }

  private mapSubscription(sub: Stripe.Subscription): Subscription {
    return {
      id: sub.id,
      providerId: sub.id,
      provider: 'stripe',
      customerId: typeof sub.customer === 'string' ? sub.customer : sub.customer.id,
      priceId: sub.items.data[0]?.price.id ?? '',
      status: this.mapSubscriptionStatus(sub.status),
      currentPeriodStart: new Date(sub.current_period_start * 1000),
      currentPeriodEnd: new Date(sub.current_period_end * 1000),
      cancelAtPeriodEnd: sub.cancel_at_period_end,
      canceledAt: sub.canceled_at ? new Date(sub.canceled_at * 1000) : undefined,
      trialStart: sub.trial_start ? new Date(sub.trial_start * 1000) : undefined,
      trialEnd: sub.trial_end ? new Date(sub.trial_end * 1000) : undefined,
      metadata: sub.metadata ?? {},
      createdAt: new Date(sub.created * 1000),
    };
  }

  private mapSubscriptionStatus(status: Stripe.Subscription.Status): SubscriptionStatus {
    switch (status) {
      case 'trialing': return 'trialing';
      case 'active': return 'active';
      case 'past_due': return 'past_due';
      case 'canceled': return 'canceled';
      case 'unpaid': return 'unpaid';
      case 'incomplete': return 'incomplete';
      default: return status as SubscriptionStatus;
    }
  }

  private mapEventType(type: string): WebhookEventType {
    const mapping: Record<string, WebhookEventType> = {
      'charge.succeeded': 'charge.succeeded',
      'charge.failed': 'charge.failed',
      'charge.refunded': 'charge.refunded',
      'customer.subscription.created': 'subscription.created',
      'customer.subscription.updated': 'subscription.updated',
      'customer.subscription.deleted': 'subscription.canceled',
      'customer.subscription.trial_will_end': 'subscription.trial_ending',
      'invoice.paid': 'invoice.paid',
      'invoice.payment_failed': 'invoice.payment_failed',
      'customer.deleted': 'customer.deleted',
    };
    return mapping[type] ?? (type as WebhookEventType);
  }

  private mapError(err: unknown): PaymentError {
    if (err instanceof PaymentError) return err;

    if (err instanceof Stripe.errors.StripeError) {
      const code = this.mapErrorCode(err.code);
      return new PaymentError(code, err.message, 'stripe', err.code);
    }

    return new PaymentError('unknown_error', String(err), 'stripe');
  }

  private mapErrorCode(code?: string): ErrorCode {
    switch (code) {
      case 'card_declined': return 'card_declined';
      case 'expired_card': return 'card_expired';
      case 'invalid_number': return 'card_invalid';
      case 'rate_limit': return 'rate_limited';
      default: return 'unknown_error';
    }
  }
}

File: sequential-1770261425547/baseline-workspace/payments/stripe/stripe.go
// Package stripe implements PaymentProvider for Stripe.
package stripe

import (
	"context"
	"time"

	"github.com/stripe/stripe-go/v76"
	"github.com/stripe/stripe-go/v76/charge"
	"github.com/stripe/stripe-go/v76/customer"
	"github.com/stripe/stripe-go/v76/paymentmethod"
	"github.com/stripe/stripe-go/v76/refund"
	"github.com/stripe/stripe-go/v76/subscription"
	"github.com/stripe/stripe-go/v76/webhook"

	"payments"
)

// Provider implements payments.PaymentProvider for Stripe.
type Provider struct {
	apiKey        string
	webhookSecret string
}

// Config holds Stripe configuration.
type Config struct {
	APIKey        string
	WebhookSecret string
}

// New creates a new Stripe provider.
func New(cfg Config) *Provider {
	stripe.Key = cfg.APIKey
	return &Provider{
		apiKey:        cfg.APIKey,
		webhookSecret: cfg.WebhookSecret,
	}
}

func (p *Provider) Name() string {
	return "stripe"
}

// ============ Charges ============

func (p *Provider) CreateCharge(ctx context.Context, req payments.ChargeRequest) (*payments.Charge, error) {
	params := &stripe.ChargeParams{
		Amount:      stripe.Int64(req.Amount),
		Currency:    stripe.String(req.Currency),
		Description: stripe.String(req.Description),
	}

	if req.CustomerID != "" {
		params.Customer = stripe.String(req.CustomerID)
	}

	if req.IdempotencyKey != "" {
		params.SetIdempotencyKey(req.IdempotencyKey)
	}

	for k, v := range req.Metadata {
		params.AddMetadata(k, v)
	}

	ch, err := charge.New(params)
	if err != nil {
		return nil, mapStripeError(err)
	}

	return mapCharge(ch), nil
}

func (p *Provider) GetCharge(ctx context.Context, chargeID string) (*payments.Charge, error) {
	ch, err := charge.Get(chargeID, nil)
	if err != nil {
		return nil, mapStripeError(err)
	}
	return mapCharge(ch), nil
}

// ============ Refunds ============

func (p *Provider) CreateRefund(ctx context.Context, req payments.RefundRequest) (*payments.Refund, error) {
	params := &stripe.RefundParams{
		Charge: stripe.String(req.ChargeID),
	}

	if req.Amount > 0 {
		params.Amount = stripe.Int64(req.Amount)
	}

	if req.Reason != "" {
		params.Reason = stripe.String(req.Reason)
	}

	if req.IdempotencyKey != "" {
		params.SetIdempotencyKey(req.IdempotencyKey)
	}

	r, err := refund.New(params)
	if err != nil {
		return nil, mapStripeError(err)
	}

	return mapRefund(r), nil
}

func (p *Provider) GetRefund(ctx context.Context, refundID string) (*payments.Refund, error) {
	r, err := refund.Get(refundID, nil)
	if err != nil {
		return nil, mapStripeError(err)
	}
	return mapRefund(r), nil
}

// ============ Customers ============

func (p *Provider) CreateCustomer(ctx context.Context, req payments.CreateCustomerRequest) (*payments.Customer, error) {
	params := &stripe.CustomerParams{
		Email: stripe.String(req.Email),
		Name:  stripe.String(req.Name),
		Phone: stripe.String(req.Phone),
	}

	for k, v := range req.Metadata {
		params.AddMetadata(k, v)
	}

	c, err := customer.New(params)
	if err != nil {
		return nil, mapStripeError(err)
	}

	return mapCustomer(c), nil
}

func (p *Provider) GetCustomer(ctx context.Context, customerID string) (*payments.Customer, error) {
	c, err := customer.Get(customerID, nil)
	if err != nil {
		return nil, mapStripeError(err)
	}
	return mapCustomer(c), nil
}

func (p *Provider) UpdateCustomer(ctx context.Context, customerID string, req payments.UpdateCustomerRequest) (*payments.Customer, error) {
	params := &stripe.CustomerParams{}

	if req.Email != nil {
		params.Email = req.Email
	}
	if req.Name != nil {
		params.Name = req.Name
	}
	if req.Phone != nil {
		params.Phone = req.Phone
	}
	for k, v := range req.Metadata {
		params.AddMetadata(k, v)
	}

	c, err := customer.Update(customerID, params)
	if err != nil {
		return nil, mapStripeError(err)
	}

	return mapCustomer(c), nil
}

func (p *Provider) DeleteCustomer(ctx context.Context, customerID string) error {
	_, err := customer.Del(customerID, nil)
	return mapStripeError(err)
}

// ============ Payment Methods ============

func (p *Provider) AttachPaymentMethod(ctx context.Context, customerID, paymentMethodID string) error {
	_, err := paymentmethod.Attach(paymentMethodID, &stripe.PaymentMethodAttachParams{
		Customer: stripe.String(customerID),
	})
	return mapStripeError(err)
}

func (p *Provider) DetachPaymentMethod(ctx context.Context, paymentMethodID string) error {
	_, err := paymentmethod.Detach(paymentMethodID, nil)
	return mapStripeError(err)
}

func (p *Provider) ListPaymentMethods(ctx context.Context, customerID string) ([]*payments.PaymentMethod, error) {
	params := &stripe.PaymentMethodListParams{
		Customer: stripe.String(customerID),
		Type:     stripe.String("card"),
	}

	iter := paymentmethod.List(params)
	var methods []*payments.PaymentMethod

	for iter.Next() {
		methods = append(methods, mapPaymentMethod(iter.PaymentMethod()))
	}

	if err := iter.Err(); err != nil {
		return nil, mapStripeError(err)
	}

	return methods, nil
}

// ============ Subscriptions ============

func (p *Provider) CreateSubscription(ctx context.Context, req payments.CreateSubscriptionRequest) (*payments.Subscription, error) {
	params := &stripe.SubscriptionParams{
		Customer: stripe.String(req.CustomerID),
		Items: []*stripe.SubscriptionItemsParams{
			{Price: stripe.String(req.PriceID)},
		},
	}

	if req.PaymentMethodID != "" {
		params.DefaultPaymentMethod = stripe.String(req.PaymentMethodID)
	}

	if req.TrialDays > 0 {
		params.TrialPeriodDays = stripe.Int64(int64(req.TrialDays))
	}

	if req.IdempotencyKey != "" {
		params.SetIdempotencyKey(req.IdempotencyKey)
	}

	for k, v := range req.Metadata {
		params.AddMetadata(k, v)
	}

	sub, err := subscription.New(params)
	if err != nil {
		return nil, mapStripeError(err)
	}

	return mapSubscription(sub), nil
}

func (p *Provider) GetSubscription(ctx context.Context, subscriptionID string) (*payments.Subscription, error) {
	sub, err := subscription.Get(subscriptionID, nil)
	if err != nil {
		return nil, mapStripeError(err)
	}
	return mapSubscription(sub), nil
}

func (p *Provider) UpdateSubscription(ctx context.Context, subscriptionID string, req payments.UpdateSubscriptionRequest) (*payments.Subscription, error) {
	params := &stripe.SubscriptionParams{}

	if req.PaymentMethodID != nil {
		params.DefaultPaymentMethod = req.PaymentMethodID
	}

	if req.CancelAtPeriodEnd != nil {
		params.CancelAtPeriodEnd = req.CancelAtPeriodEnd
	}

	for k, v := range req.Metadata {
		params.AddMetadata(k, v)
	}

	// Plan change requires updating subscription items
	if req.PriceID != nil {
		sub, _ := subscription.Get(subscriptionID, nil)
		if sub != nil && len(sub.Items.Data) > 0 {
			params.Items = []*stripe.SubscriptionItemsParams{
				{
					ID:    stripe.String(sub.Items.Data[0].ID),
					Price: req.PriceID,
				},
			}
		}
	}

	sub, err := subscription.Update(subscriptionID, params)
	if err != nil {
		return nil, mapStripeError(err)
	}

	return mapSubscription(sub), nil
}

func (p *Provider) CancelSubscription(ctx context.Context, subscriptionID string, req payments.CancelSubscriptionRequest) (*payments.Subscription, error) {
	if req.CancelAtPeriodEnd {
		// Cancel at period end
		sub, err := subscription.Update(subscriptionID, &stripe.SubscriptionParams{
			CancelAtPeriodEnd: stripe.Bool(true),
		})
		if err != nil {
			return nil, mapStripeError(err)
		}
		return mapSubscription(sub), nil
	}

	// Immediate cancellation
	sub, err := subscription.Cancel(subscriptionID, nil)
	if err != nil {
		return nil, mapStripeError(err)
	}

	return mapSubscription(sub), nil
}

func (p *Provider) ListSubscriptions(ctx context.Context, customerID string) ([]*payments.Subscription, error) {
	params := &stripe.SubscriptionListParams{
		Customer: customerID,
	}

	iter := subscription.List(params)
	var subs []*payments.Subscription

	for iter.Next() {
		subs = append(subs, mapSubscription(iter.Subscription()))
	}

	if err := iter.Err(); err != nil {
		return nil, mapStripeError(err)
	}

	return subs, nil
}

// ============ Webhooks ============

func (p *Provider) ParseWebhook(payload []byte, signature string) (*payments.WebhookEvent, error) {
	event, err := webhook.ConstructEvent(payload, signature, p.webhookSecret)
	if err != nil {
		return nil, payments.NewPaymentError(
			payments.ErrInvalidWebhook,
			"Invalid webhook signature",
			"stripe",
		)
	}

	return &payments.WebhookEvent{
		ID:        event.ID,
		Type:      mapEventType(event.Type),
		Provider:  "stripe",
		Data:      event.Data.Object,
		CreatedAt: time.Unix(event.Created, 0),
	}, nil
}

// ============ Mappers ============

func mapCharge(ch *stripe.Charge) *payments.Charge {
	return &payments.Charge{
		ID:             ch.ID,
		ProviderID:     ch.ID,
		Provider:       "stripe",
		Amount:         ch.Amount,
		AmountRefunded: ch.AmountRefunded,
		Currency:       string(ch.Currency),
		Status:         mapChargeStatus(ch.Status),
		CustomerID:     ch.Customer.ID,
		Description:    ch.Description,
		FailureCode:    string(ch.FailureCode),
		FailureReason:  ch.FailureMessage,
		Metadata:       ch.Metadata,
		CreatedAt:      time.Unix(ch.Created, 0),
	}
}

func mapChargeStatus(status string) payments.ChargeStatus {
	switch status {
	case "succeeded":
		return payments.ChargeStatusSucceeded
	case "pending":
		return payments.ChargeStatusPending
	case "failed":
		return payments.ChargeStatusFailed
	default:
		return payments.ChargeStatus(status)
	}
}

func mapRefund(r *stripe.Refund) *payments.Refund {
	return &payments.Refund{
		ID:         r.ID,
		ProviderID: r.ID,
		Provider:   "stripe",
		ChargeID:   r.Charge.ID,
		Amount:     r.Amount,
		Currency:   string(r.Currency),
		Status:     mapRefundStatus(r.Status),
		Reason:     string(r.Reason),
		CreatedAt:  time.Unix(r.Created, 0),
	}
}

func mapRefundStatus(status stripe.RefundStatus) payments.RefundStatus {
	switch status {
	case stripe.RefundStatusSucceeded:
		return payments.RefundStatusSucceeded
	case stripe.RefundStatusPending:
		return payments.RefundStatusPending
	case stripe.RefundStatusFailed:
		return payments.RefundStatusFailed
	default:
		return payments.RefundStatus(status)
	}
}

func mapCustomer(c *stripe.Customer) *payments.Customer {
	return &payments.Customer{
		ID:         c.ID,
		ProviderID: c.ID,
		Provider:   "stripe",
		Email:      c.Email,
		Name:       c.Name,
		Phone:      c.Phone,
		Metadata:   c.Metadata,
		CreatedAt:  time.Unix(c.Created, 0),
	}
}

func mapPaymentMethod(pm *stripe.PaymentMethod) *payments.PaymentMethod {
	method := &payments.PaymentMethod{
		ID:         pm.ID,
		ProviderID: pm.ID,
		Provider:   "stripe",
		Type:       payments.PaymentMethodType(pm.Type),
		CreatedAt:  time.Unix(pm.Created, 0),
	}

	if pm.Card != nil {
		method.Card = &payments.CardDetails{
			Brand:    string(pm.Card.Brand),
			Last4:    pm.Card.Last4,
			ExpMonth: int(pm.Card.ExpMonth),
			ExpYear:  int(pm.Card.ExpYear),
		}
	}

	return method
}

func mapSubscription(sub *stripe.Subscription) *payments.Subscription {
	s := &payments.Subscription{
		ID:                 sub.ID,
		ProviderID:         sub.ID,
		Provider:           "stripe",
		CustomerID:         sub.Customer.ID,
		Status:             mapSubscriptionStatus(sub.Status),
		CurrentPeriodStart: time.Unix(sub.CurrentPeriodStart, 0),
		CurrentPeriodEnd:   time.Unix(sub.CurrentPeriodEnd, 0),
		CancelAtPeriodEnd:  sub.CancelAtPeriodEnd,
		Metadata:           sub.Metadata,
		CreatedAt:          time.Unix(sub.Created, 0),
	}

	if len(sub.Items.Data) > 0 {
		s.PriceID = sub.Items.Data[0].Price.ID
	}

	if sub.CanceledAt > 0 {
		t := time.Unix(sub.CanceledAt, 0)
		s.CanceledAt = &t
	}

	if sub.TrialStart > 0 {
		t := time.Unix(sub.TrialStart, 0)
		s.TrialStart = &t
	}

	if sub.TrialEnd > 0 {
		t := time.Unix(sub.TrialEnd, 0)
		s.TrialEnd = &t
	}

	return s
}

func mapSubscriptionStatus(status stripe.SubscriptionStatus) payments.SubscriptionStatus {
	switch status {
	case stripe.SubscriptionStatusTrialing:
		return payments.SubscriptionStatusTrialing
	case stripe.SubscriptionStatusActive:
		return payments.SubscriptionStatusActive
	case stripe.SubscriptionStatusPastDue:
		return payments.SubscriptionStatusPastDue
	case stripe.SubscriptionStatusCanceled:
		return payments.SubscriptionStatusCanceled
	case stripe.SubscriptionStatusUnpaid:
		return payments.SubscriptionStatusUnpaid
	case stripe.SubscriptionStatusIncomplete:
		return payments.SubscriptionStatusIncomplete
	default:
		return payments.SubscriptionStatus(status)
	}
}

func mapEventType(t string) payments.WebhookEventType {
	switch t {
	case "charge.succeeded":
		return payments.WebhookEventChargeSucceeded
	case "charge.failed":
		return payments.WebhookEventChargeFailed
	case "charge.refunded":
		return payments.WebhookEventChargeRefunded
	case "customer.subscription.created":
		return payments.WebhookEventSubscriptionCreated
	case "customer.subscription.updated":
		return payments.WebhookEventSubscriptionUpdated
	case "customer.subscription.deleted":
		return payments.WebhookEventSubscriptionCanceled
	case "customer.subscription.trial_will_end":
		return payments.WebhookEventSubscriptionTrialEnd
	case "invoice.paid":
		return payments.WebhookEventInvoicePaid
	case "invoice.payment_failed":
		return payments.WebhookEventInvoicePaymentFailed
	default:
		return payments.WebhookEventType(t)
	}
}

func mapStripeError(err error) error {
	if err == nil {
		return nil
	}

	stripeErr, ok := err.(*stripe.Error)
	if !ok {
		return payments.NewPaymentError(payments.ErrUnknown, err.Error(), "stripe")
	}

	var code payments.ErrorCode
	switch stripeErr.Code {
	case stripe.ErrorCodeCardDeclined:
		code = payments.ErrCardDeclined
	case stripe.ErrorCodeExpiredCard:
		code = payments.ErrCardExpired
	case stripe.ErrorCodeInvalidNumber:
		code = payments.ErrCardInvalid
	case stripe.ErrorCodeRateLimitExceeded:
		code = payments.ErrRateLimited
	default:
		code = payments.ErrUnknown
	}

	paymentErr := payments.NewPaymentError(code, stripeErr.Msg, "stripe")
	paymentErr.ProviderCode = string(stripeErr.Code)
	return paymentErr
}
